### PR TITLE
[Re-landing 68111] Add JIT graph fuser for oneDNN Graph API (Preview4.1)

### PR DESCRIPTION
**Pull request by chunyuan-w**
Re-landing https://github.com/pytorch/pytorch/pull/68111

## Description
Preview4 PR of this [RFC](https://github.com/pytorch/pytorch/issues/49444).

On the basis of https://github.com/pytorch/pytorch/pull/50256, the below improvements are included:

- The [preview4 release branch](https://github.com/oneapi-src/oneDNN/releases/tag/graph-v0.4.1) of the oneDNN Graph API is used
- The fuser now works with the profiling graph executor. We have inserted type check nodes to guard the profiled tensor properties.

### User API:
The optimization pass is disabled by default. Users could enable it by:
```
torch.jit.enable_onednn_fusion(True)
```

### Performance:
[pytorch/benchmark](https://github.com/pytorch/benchmark) tool is used to compare the performance:
- SkyLake 8180 (1 socket of 28 cores):

  image

- SkyLake 8180 (single thread):

  image
 \* By mapping hardswish to oneDNN Graph, it’s 8% faster than PyTorch JIT (NNC + OFI)
  \** We expect performance gain after mapping transpose, contiguous & view to oneDNN graph ops

### Directory structure of the integration code
Fuser-related code are placed under:
```
torch/csrc/jit/codegen/onednn/
```

Optimization pass registration is done in:
```
torch/csrc/jit/passes/onednn_graph_fuser.h
```

CMake for the integration code is:
```
caffe2/CMakeLists.txt
```

## Limitations

- In this PR, we have only supported the optimization on Linux platform. The support on Windows and MacOS will be enabled as the next step.
- We have only optimized the inference use case.

### CODE REVIEW

**Comment by sanchitintel** [comment]
Failures are unrelated.

**Comment by sanchitintel** [comment]
Hello @malfet, we created this PR for oneDNN Graph Preview4, and closed the other one (#74572).
Sorry for the inconvenience!
BTW, as per [your advice](https://github.com/pytorch/pytorch/pull/74572#issuecomment-1075645407), I had cherry-picked the [landed commit](https://github.com/pytorch/pytorch/commit/cd17683aa7d9c0947df45a1ab53627feff795587) before applying any other changes. Thank you!

**Review by malfet — CHANGES_REQUESTED** [review]
In my opinion, a good solution to the problem would be to create `BUILD_WITH_LLGA` option, which would be conditional whatever condition one chooses, and in `torch/csrc/jit/python/init.cpp` defined the callbacks only if `BUILD_WITH_LLGA` is enabled

**Comment by sanchitintel** [comment]
The CI failure is unrelated. Will rebase to re-run CI when the ongoing CI issues would be fixed. Thanks!

**Comment by sanchitintel** [comment]
Hi @malfet, if you'd like, I can make `BUILD_ONEDNN_GRAPH` a `cmake_dependent_option` instead, as it looks like `MKLDNN_FOUND` would always be true when `USE_MKLDNN` is true (if oneDNN repo can't be found, then `find_mkldnn.cmake` creates it by pulling it). That'd also increase code-clarity.

Thanks!

**Comment by malfet** [comment]
@pytorch label ciflow/trunk

File: CMakeLists.txt [hunk 1]
```diff
@@ -688,8 +688,14 @@ if(USE_FBGEMM AND ((CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64" AND CMAKE_SIZEOF_VO
   set(USE_FBGEMM OFF)
 endif()
 
+set(BUILD_ONEDNN_GRAPH OFF)
+
 include(cmake/Dependencies.cmake)
 
+if(BUILD_ONEDNN_GRAPH)
+  string(APPEND CMAKE_CXX_FLAGS " -DBUILD_ONEDNN_GRAPH")
```
Comment (malfet) [inline CMakeLists.txt:696]: Actually, this is less then ideal, as it passes this option to all libraries, rather than to just torch_cpu and torch_python

Comment (sanchitintel) [reply CMakeLists.txt:696]: Thanks for your advice, @malfet! I'd try to ensure that it'd only be passed to the libraries that use it.
However, the same thing holds true for other flags in `CMakeLists.txt` as well, for example, https://github.com/pytorch/pytorch/blob/master/CMakeLists.txt#L703-L742.

Comment (sanchitintel) [reply CMakeLists.txt]: @malfet, I've made this change. Thank you! :)
Can you please re-import the PR? Sorry for the inconvenience!

**Comment by sanchitintel** [comment]
Looks like the CI system is down. 😅 

image

**Comment by sanchitintel** [comment]
The CI failure is unrelated (it's related to CI). Thanks!
<details>
image
</details>

**Review by malfet — APPROVED** [review]

**Comment by Jianhui-Li** [comment]
@malfet could you please land the PR as it passes all the CI tests? The 2 failed cases seem irrelevant.

**Comment by sanchitintel** [comment]
`test_demux_mux_datapipe` is failing for other PRs as well.
Apart from that test failing for many CI checks, the `rocm` failure is also unrelated.

**Comment by sanchitintel** [comment]
The 4 failing CI checks had failed due to unrelated reasons. Thanks!

**Comment by sanchitintel** [comment]
Hi @clee2000 @seemethere, 

`do_merge` for this PR failed with an `xla` related commit. Not sure but maybe that's because another commit ID would've to be added here - https://github.com/pytorch/pytorch/blob/master/.github/scripts/gitutils.py#L198-L207. Please advise. Thank you!

image

**Comment by sanchitintel** [comment]
@pytorch label release notes: jit

**Comment by janeyx99** [comment]
@pytorchbot revert this please

Sorry, reverting as this caused macos test failures on trunk https://hud.pytorch.org/minihud?name_filter=trunk%20/%20macos-11-py3-x86-64%20/%20test%20(default,%201,%202,%20macos-11). Please include the ciflow/trunk label on your next reland attempt!

**Comment by janeyx99** [comment]
> Revert unsuccessful: please retry the command explaining why the revert is necessary, e.g. @pytorchbot revert this as it breaks mac tests on trunk, see .

This is unintended behavior--fixing bot with https://github.com/pytorch/test-infra/pull/311

**Comment by sanchitintel** [comment]
Hi @janeyx99, the macos failures seem to be unrelated because oneDNN Graph is not built with macOS.

**Comment by osalpekar** [comment]
Co-Dev Tooling should not try to revert a PR that's already reverted (though there's no negative side-effect since the revert will just fail). We can check the Reverted label on the PR for for this, filing task for this

**Comment by bigfootjon** [comment]
There's actually 2 bugs here: co-development isn't enabled for pytorch/pytorch right now lol. This appears to be a pytorch-canary gone rogue :(

I'll fix both on Monday

**Pull request closed without merging**